### DISCUSSION TITLE
Best Practices for Dependency Injection in Symfony Controllers: Constructor vs. Method Arguments

### DISCUSSION THREAD

**Discussion by nietonchique** [symfony/symfony#53465] [Q&A]
Hello,

I'm seeking conceptual insights (not performance-related) on dependency injection in Symfony controllers. There are two prevalent approaches: injecting dependencies via the constructor and through method arguments defining the endpoint. Utilizing constructor injection consistently results in slightly lower RPS (e.g., approximately half in a HelloWorld scenario), but the difference is marginal (in microseconds).

Question: From a design perspective, should dependency injection be exclusively through the constructor, reserving method arguments solely for incoming data? Alternatively, is it preferable to mix both approaches, or shift entirely to injecting dependencies via method arguments?

Your insights and experiences would be highly valued.

Thank you!

**Comment by yceruto** [comment]
I think there is no right way for this answer, but it's important to be consistent if you decide on one approach or another.

Personally, I like to inject all dependencies using the constructor and leave the method arguments only for request parameters. Mixing them into the method arguments feels kind of messy to me, mainly if there are many dependencies.

If you like to structure or split your controllers into single-action classes, it would make less difference, but still, I'd prefer the constructor injection approach.

**Reply by GromNaN** [reply]
The Symfony demo uses mainly argument injection in the controllers: https://github.com/symfony/demo/blob/v2.5.0/src/Controller/Admin/BlogController.php

**Reply by yceruto** [reply]
Yes, it's apparently better in that case because it defines all controllers in one class, so you inject the service into the method/controller where it's really necessary. It's not wrong, but just a different approach.

In my opinion, dependencies should _always_ be injected through constructors for controllers to make their meaning clearer. Mixing them with request parameters might be confusing for new developers, as it becomes unclear where they come from (from the request or from DI?). Having these two groups separated would avoid that confusion.

**Comment by GromNaN** [comment]
I wondered if there was a difference in performance, but probably not. Dependency injections in the constructor are resolved at compile time (by autowiring or explicit configuration). Same for the [`ServiceValueResolver`](https://github.com/symfony/symfony/blob/7.0/src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php) that is optimized by the [`RegisterControllerArgumentLocatorsPass`](https://github.com/symfony/symfony/blob/7.0/src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php).
